velox/functions/prestosql/DecimalArithmetic.cpp

```diff
@@ -0,0 +1,247 @@
+        args[0]->encoding() == VectorEncoding::Simple::FLAT &&
+        args[1]->isConstantEncoding()) {
+      auto flatValues = args[0]->asUnchecked<FlatVector<A>>();
+      auto constant = args[1]->asUnchecked<SimpleVector<B>>()->valueAt(0);
+      auto rawValues = flatValues->mutableRawValues();
+      auto rawResults =
+          getRawResults(rows, args[1], resultType, context, result);
+      context->applyToSelectedNoThrow(rows, [&](auto row) {
+        Operation::template apply<R, A, B>(
+            rawResults[row], rawValues[row], constant, aRescale, bRescale);
+      });
+    } else if (
+        args[0]->encoding() == VectorEncoding::Simple::FLAT &&
+        args[1]->encoding() == VectorEncoding::Simple::FLAT) {
+      auto flatA = args[0]->asUnchecked<FlatVector<A>>();
+      auto rawA = flatA->mutableRawValues();
+      auto flatB = args[1]->asUnchecked<FlatVector<B>>();
+      auto rawB = flatB->mutableRawValues();
+
+      R* rawResults = nullptr;
+      if (!(*result)) {
+        if (resultType->kind() == flatA->typeKind() &&
+            BaseVector::isReusableFlatVector(args[0])) {
+          rawResults = (R*)rawA;
+          *result = std::move(args[0]);
+        } else if (
+            resultType->kind() == flatB->typeKind() &&
+            BaseVector::isReusableFlatVector(args[1])) {
+          rawResults = (R*)rawB;
+          *result = std::move(args[1]);
+        }
+      }
+      if (!rawResults) {
+        rawResults = prepareResults(rows, resultType, context, result);
+      }
+      context->applyToSelectedNoThrow(rows, [&](auto row) {
+        Operation::template apply<R, A, B>(
+            rawResults[row], rawA[row], rawB[row], aRescale, bRescale);
+      });
+    } else {
+      exec::DecodedArgs decodedArgs(rows, args, context);
+      auto a = decodedArgs.at(0);
+      auto b = decodedArgs.at(1);
+      BaseVector::ensureWritable(rows, resultType, context->pool(), result);
+      auto rawResults =
+          (*result)->asUnchecked<FlatVector<R>>()->mutableRawValues();
+      context->applyToSelectedNoThrow(rows, [&](auto row) {
+        Operation::template apply<R, A, B>(
+            rawResults[row],
+            a->valueAt<A>(row),
+            b->valueAt<B>(row),
+            aRescale,
+            bRescale);
+      });
+    }
+  }
+
+ private:
+  R* prepareResults(
+      const SelectivityVector& rows,
+      const TypePtr& resultType,
+      exec::EvalCtx* context,
+      VectorPtr* result) const {
+    BaseVector::ensureWritable(rows, resultType, context->pool(), result);
+    (*result)->clearNulls(rows);
+    return (*result)->asUnchecked<FlatVector<R>>()->mutableRawValues();
+  }
+
+  R* getRawResults(
+      const SelectivityVector& rows,
+      VectorPtr& flat,
+      const TypePtr& resultType,
+      exec::EvalCtx* context,
+      VectorPtr* result) const {
+    // Check if input can be reused for results.
+    R* rawResults;
+    if (!(*result) && flat->typeKind() == resultType->kind() &&
```

Comment:
It is not enough to check that type kinds match. The full types must match, including precision and scale.

Reply:
Checking TypeKind should be sufficient. We have to just make sure that a LongDecimal raw buffer to another LongDecimal pointer. LongDecimal or ShortDecimal don't hold the precision and scale.

Reply:
Actually, we must ensure that the vector has the right type, not just right type kind. We are going to use it for results and if types don't match the result vector will be of the wrong type.

Reply:
For most of the arithmetic operations the result's precision and scale will be different from argument's precision and scale. Would you please give an example when does the flatvector of one TypeKind has a different underlying type? I thought there is a tight binding between TypeKind and the native type.

Reply:
I think Masha's point is that the result vector type (TypePtr) should not change. I guess we cannot reuse for decimals.

Reply:
Another option is to change the resultType.

Reply:
> I think Masha's point is that the result vector type (TypePtr) should not change. I guess we cannot reuse for decimal arithmetic.

That's what I was trying to say. Currently vector's type is specified at construction and cannot be changed, hence, reuse is only possible if types match exactly.  Maybe we could consider a way to allow to re-using the vector and changing the type, but that would require separate thinking, designing and implementing.